src/slacko.ml

```diff
@@ -765,33 +765,22 @@ let groups_list ?exclude_archived session =
       (match d |> groups_list_obj_of_yojson with
         | Result.Ok x -> `Success x.groups
         | Result.Error x -> `ParseFailure x)
-    | #bot_error
     | #parsed_auth_error as res -> res
     | _ -> `Unknown_error
 
-exception No_matches
-exception No_unique_matches
-exception Lookup_failed
-
 (* look up the id of query from results provided by the listfn *)
 let lookupk session listfn filterfn k =
   match%lwt listfn session with
-  | `Success channels -> (match List.filter filterfn channels with
-    | [] -> Lwt.fail No_matches
-    | [x] -> Lwt.return @@ k x
-    | _ -> Lwt.fail No_unique_matches)
-  | _ -> Lwt.fail Lookup_failed
+  | #parsed_auth_error as e -> Lwt.return e
+  | `Success items -> Lwt.return @@ k @@ List.filter filterfn items
```

Comment:
I bounced back and forth a few times on the "exactly one result" thing. While it belongs in `lookupk` conceptually, putting it there just adds a pile of extra values to pass around. If we handle "too many" and "too few" cases separately, we get this:
```ocaml
let lookupk session listfn filterfn k =
  match%lwt listfn session with
  | #parsed_auth_error as e -> Lwt.return e
  | `Success items -> match List.filter filterfn items with
    | [item] -> Lwt.return @@ k (`Found item)
    | [] -> Lwt.return @@ k (`Too_few)
    | _::_::_ -> Lwt.return @@ k (`Too_many)

let id_of_channel session = function
  | ChannelId id -> Lwt.return @@ `Found id
  | ChannelName name ->
    let base = String.sub name 1 @@ String.length name - 1 in
    lookupk session channels_list (fun (x:channel_obj) -> x.name = base) @@ function
    | `Found {id = ChannelId s; _} -> `Found s
    | `Found {id = ChannelName _; _} -> `Channel_not_found
    | `Too_few -> `Channel_not_found
    | `Too_many -> `Channel_not_found
```

We could reduce this to a pair of "found"/"not found" cases and pass a `'b option` to `k`, but we'd still be splitting the result into multiple parts and then recombining them in `k`. We could also return a "not found" error directly from `lookupk`, but then we'd need to pass that in (because it's different for different callers) and we'd still need to handle the `<thing>Id` vs `<thing>Name` split in the caller anyway.

I suppose we could merge `filterfn` and `k` into a single function that returns `Some s` for a matching object with an appropriate id and `None` for a mismatch or no id, but we'd still have to pass in the "not found value. I'll try that and see if it ends up being cleaner.